src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -174,40 +174,26 @@ public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, lon
             // Allocate an active request
             QuicStream? quicStream = null;
             Http3RequestStream? requestStream = null;
-            ValueTask waitTask = default;
 
             try
             {
                 try
                 {
-                    while (true)
+                    if (_connection != null)
                     {
-                        lock (SyncObj)
+                        if (HttpTelemetry.Log.IsEnabled() && queueStartingTimestamp == 0 && _connection.GetRemoteAvailableBidirectionalStreamCount() == 0)
```

Comment:
`GetRemoteAvailableBidirectionalStreamCount` is a native call. I suppose it will be faster to always capture the time and always issue a telemetry event (if telemetry is enabled) rather than put a native call in a condition. Thoughts @ManickaP @MihaZupan ?

Reply:
I'll just confirm that the call means interop (that's why it's not just property but a method) which does incur some CPU cycles.
Also since this is outside of lock now, it can be called from multiple threads at the same time (multiple parallel `HttpClient.SendAsync`), it's not 100% reliable.

Reply:
Have we also considered copying the data to the managed side (and keep them updated via the STREAMS_AVAILABLE event)?

Reply:
If we decide to keep our own stream counting, we'll probably do it on `MsQuicConnection` side. But I'd rather leave that outside of this PR. Either way, we should file (or find an existing and update) an issue with support for multiple H/3 connections where this can discussed.

Reply:
I haven't been following all the changes in this PR, but is this comment correct?
https://github.com/dotnet/runtime/blob/2c299dcdadfe982e0a55a168d5ad17b2ccc29982/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs#L187

Isn't the common case that there are streams available, and the call will return synchronously?

> Since in HTTP2, we log all telemetry

We won't log `RequestLeftQueue` if a connection is already available. We only do it if we had to block until a stream is available.


I'm not sure I understand why the telemetry here has to change. My naive understanding is that we're just replacing `WaitForAvailableBidirectionalStreamsAsync` with `OpenBidirectionalStreamAsync`.

Reply:
> the call will return synchronously

The call will never return synchronously, the underlying msquic always open the streams asynchronously, even if available.

> I'm not sure I understand why the telemetry here has to change. My naive understanding is that we're just replacing WaitForAvailableBidirectionalStreamsAsync with OpenBidirectionalStreamAsync.

While previously a synchronous completion was possible, since updating to msquic 2.0 it is not. Plus, the whole loop with locks, and native calls to GetRemoteAvailableBidirectionalStreamCount, and WaitForAvailableBidirectionalStreamsAsync showed bad performance, so we are now delegating the wait to msquic.

I would say, there's no 100% reliable way to say whether the call would end up waiting or not, even with checking GetRemoteAvailableBidirectionalStreamCount (without a lock around). The stream IDs might be available and then gone to different streams in different threads; and when 0 is returned, a stream might get back available in a moment.

Reply:
Ah, I see... I was reading the `OpenBidirectionalStreamAsync` implementation of `MockConnection` 🤦‍♂️ 

If it's not cheap to determine whether a stream will be immediately available, I'm fine with always logging `RequestLeftQueue`.

In this case, I think the logic around `queueStartingTimestamp` should be simplified. We can delete:
- [these lines](https://github.com/dotnet/runtime/blob/2c299dcdadfe982e0a55a168d5ad17b2ccc29982/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs#L185-L189) and
- [these lines](https://github.com/dotnet/runtime/blob/2c299dcdadfe982e0a55a168d5ad17b2ccc29982/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L947-L951)